external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -141,83 +130,676 @@ struct DependencyTOML<'a>(PM::PackageName, &'a Dependency);
+            }
+        }
+        Ok(None)
     }
 
-    /// Create a dependency graph by reading a lock file.
+    /// Given a dependency in the parent's manifest file, creates a sub-graph for this dependency.
+    fn new_for_dep<Progress: Write>(
+        parent: &PM::DependencyKind,
+        dep: &PM::Dependency,
+        mode: DependencyMode,
+        parent_pkg: PM::PackageName,
+        dep_pkg_name: PM::PackageName,
+        dep_pkg_path: PathBuf,
+        internal_dependencies: &mut VecDeque<(PM::PackageName, PM::InternalDependency)>,
+        dependency_cache: &mut DependencyCache,
+        progress_output: &mut Progress,
+    ) -> Result<(DependencyGraph, bool)> {
+        let (pkg_graph, is_override) = match dep {
+            PM::Dependency::Internal(d) => {
+                DependencyGraph::check_for_dep_cycles(
+                    d.clone(),
+                    dep_pkg_name,
+                    internal_dependencies,
+                )?;
+                dependency_cache
+                    .download_and_update_if_remote(dep_pkg_name, &d.kind, progress_output)
+                    .with_context(|| format!("Fetching '{}'", dep_pkg_name))?;
+                let pkg_path = dep_pkg_path.join(local_path(&d.kind));
+                let manifest_string =
+                    std::fs::read_to_string(pkg_path.join(SourcePackageLayout::Manifest.path()))
+                        .with_context(|| format!("Parsing manifest for '{}'", dep_pkg_name))?;
+                let lock_string =
+                    std::fs::read_to_string(pkg_path.join(SourcePackageLayout::Lock.path())).ok();
+                // save dependency for cycle detection
+                internal_dependencies.push_front((dep_pkg_name, d.clone()));
+                let (mut pkg_graph, _) = DependencyGraph::get(
+                    &d.kind,
+                    pkg_path.clone(),
+                    manifest_string,
+                    lock_string,
+                    internal_dependencies,
+                    dependency_cache,
+                    progress_output,
+                )?;
+                internal_dependencies.pop_front();
+                // reroot all packages to normalize local paths across all graphs
+                for (_, p) in pkg_graph.package_table.iter_mut() {
+                    p.kind.reroot(parent)?;
+                }
+                (pkg_graph, d.dep_override)
+            }
+            PM::Dependency::External(resolver) => {
+                let pkg_graph = DependencyGraph::get_external(
+                    mode,
+                    parent_pkg,
+                    dep_pkg_name,
+                    *resolver,
+                    &dep_pkg_path,
+                    progress_output,
+                )?;
+                (pkg_graph, false)
+            }
+        };
+        Ok((pkg_graph, is_override))
+    }
+
+    /// Cycle detection to avoid infinite recursion due to the way we construct internally resolved
```

Comment:
Remind me -- where does the infinite recursion come from in the current implementation? I was hopeful that this simplification of the implementation would get rid of that, but I hadn't thought it all the way through.

Reply:
The algorithm starts by looking at direct dependencies of the `Root` package, and if lock files for these are available (and up-to-date) then we are done. If they are not, we recursively look at dependencies of each of the `Root` package's dependencies and so on, until we (hopefully) get to the "leaf" dependencies that have no dependencies of their own. When that happens, the recursion ends, we generated package structs for leaves and start unwinding recursion, merging smaller graphs into larger ones, until all is merged into root. 

A wrinkle here is a lack of true "leaf" dependencies by them cycling back to other dependencies in the graph, in which case the recursion will never end (as we don't do merging or individual package comparison/resolution at this point).

Reply:
Yeah, makes sense, and sadly unavoidable!  Thanks for explaining.